ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/AtomicActionRecoveryModule.java

```diff
@@ -223,17 +229,42 @@ private void processTransactionsStatus()
                 try {
                     if (_recoveryStore.currentState(currentUid,
                             _transactionType) != StateStatus.OS_UNKNOWN) {
-                        doRecoverTransaction(currentUid);
+                        RecoverAtomicAction rcvAtomicAction = doRecoverTransaction(currentUid);
+
+                        if (rcvAtomicAction.hasPreparedParticipants() || rcvAtomicAction.hasFailedParticipants() ||
+                                rcvAtomicAction.hasHeuristicParticipants()) {
+                            /*
+                             * If some participants in this AtomicAction fail to commit while others complete with a
+                             * heuristic outcome, another recovery attempt should be made to allow the failed participants
+                             * to fulfill their intention.
+                             */
+                            if (rcvAtomicAction.hasHeuristicParticipants() && !rcvAtomicAction.hasFailedParticipants()) {
```

Comment:
I am not sure if your reason for considering prepared participants is the same I am now of the mind we seem like we need to do this (before I thought checking failed participants, and heuristic ones, would be enough) but I am getting to be of the mind it is possible for a recovered transaction to have something in the prepared list now. It's quite complicated but I think if a BasicAction commits, and the first participant results in a heuristic rollback then the BasicAction will move to aborting at https://github.com/jbosstm/narayana/blob/9da7020ce86d109fe997cb8006ff8c3d95a71a08/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/BasicAction.java#L2967 and then (I suspect) if the other participants can't be contacted then when the transaction recovers then it will be in a status to do a rollback through the RecoverAtomicAction::replayPhase2 (rather than a commit). The difference of commit and rollback results in  why preparedList not being empty becomes important because when we try to commit AbstractRecords (and say they can't be contacted) I expect they go in the failed list, but I think if we try to rollback resources and fail temporarily they just seem to stay in prepared list.

So given all that (and I know that's a lot of pre-amble) I think this condition here needs to change. I think it needs to be: `rcvAtomicAction.hasHeuristicParticipants() && !rcvAtomicAction.hasFailedParticipants() && !rcvAtomicAction.hasPreparedParticipants()`

I am not certain so please do review my comment and the relevant code and see if you agree.

Reply:
Thank you. This is resolved to my satisfaction in the latest code base.

Reply:
With the new commit, I was able to simplify this `if statement` to something simpler. Thanks